### PR TITLE
[client]: add timeouts to all request kinds + default timeout

### PR DESCRIPTION
**Pull request by niklasad1**
Add default timeout on the all the clients requests to 60 seconds.

The PR itself is !DRY of  in a few places.

### CODE REVIEW

File: http-client/src/client.rs [hunk 1]
```diff
@@ -25,17 +28,25 @@ impl HttpClientBuilder {
 		self
 	}
 
+	/// Set request timeout (default is 60 seconds).
+	///
+	/// None - implies that no timeout is used.
+	pub fn request_timeout(mut self, timeout: Option<Duration>) -> Self {
```
Comment (niklasad1) [inline http-client/src/client.rs:34]: little annoying with option here.

File: http-client/src/client.rs [hunk 2]
```diff
@@ -46,16 +57,20 @@ pub struct HttpClient {
...
 	async fn notification<'a>(&self, method: &'a str, params: JsonRpcParams<'a>) -> Result<(), Error> {
 		let notif = JsonRpcNotificationSer::new(method, params);
-		self.transport
-			.send(serde_json::to_string(&notif).map_err(Error::ParseError)?)
-			.await
-			.map_err(|e| Error::Transport(Box::new(e)))
+		let fut = self.transport.send(serde_json::to_string(&notif).map_err(Error::ParseError)?);
+		match call_with_maybe_timeout(fut, self.request_timeout).await {
+			Ok(Ok(ok)) => Ok(ok),
+			Err(_) => Err(Error::RequestTimeout),
```
Comment (niklasad1) [inline http-client/src/client.rs:71]: this is `tokio::time::Elapsed`

File: ws-client/src/client.rs [hunk 3]
```diff
@@ -343,19 +359,10 @@ impl Client for WsClient {
...
-			send_back_rx.await
-		};
+		let res = crate::helpers::call_with_maybe_timeout(send_back_rx, self.request_timeout).await;
 
 		self.id_guard.reclaim_request_id();
-		let json_value = match send_back_rx_out {
+		let json_value = match res {
 			Ok(Ok(v)) => v,
 			Ok(Err(err)) => return Err(err),
 			Err(_) => return Err(self.read_error_from_backend().await),
```
Comment (niklasad1) [inline ws-client/src/client.rs:368]: this is `oneshot::Cancled err` in that case the background thread was terminated and we read the error message from the backend.

File: ws-client/src/client.rs [hunk 4]
```diff
@@ -312,7 +314,21 @@ impl Client for WsClient {
 			Error::ParseError(e)
 		})?;
 		log::trace!("[frontend]: send notification: {:?}", raw);
-		let res = self.to_back.clone().send(FrontToBack::Notification(raw)).await;
+
+		let mut sender = self.to_back.clone();
+		let fut = sender.send(FrontToBack::Notification(raw));
+
+		let res = if let Some(dur) = self.request_timeout {
```
Comment (niklasad1) [inline ws-client/src/client.rs:321]: this annoying of boiler plate but I found no good way of having a helper for `Result<T>`,  and `T` when the receiver is either expects `T` or `Result<T>`

on option is to use generic as the `http client` is using with `F: Fut -> Result<Fut::Output, Error>` but that became super complicated with lots of nested results.

File: http-client/src/client.rs (hunk 1 above)
Comment (maciejhirsz) [reply http-client/src/client.rs:34]: I reckon that's fine.

**Review by maciejhirsz — APPROVED** [review]
LGTM, small nits.

File: http-client/src/client.rs [hunk 5]
```diff
@@ -46,16 +57,20 @@ pub struct HttpClient {
...
 #[async_trait]
 impl Client for HttpClient {
 	async fn notification<'a>(&self, method: &'a str, params: JsonRpcParams<'a>) -> Result<(), Error> {
 		let notif = JsonRpcNotificationSer::new(method, params);
-		self.transport
-			.send(serde_json::to_string(&notif).map_err(Error::ParseError)?)
-			.await
-			.map_err(|e| Error::Transport(Box::new(e)))
+		let fut = self.transport.send(serde_json::to_string(&notif).map_err(Error::ParseError)?);
+		match call_with_maybe_timeout(fut, self.request_timeout).await {
```
Comment (dvdplm) [inline http-client/src/client.rs:69]: I think we can drop the `maybe` here; `call_with_timeout` is verbose enough! :)

File: http-client/src/client.rs [hunk 6]
```diff
@@ -46,16 +57,20 @@ pub struct HttpClient {
 	transport: HttpTransportClient,
 	/// Request ID that wraps around when overflowing.
 	request_id: AtomicU64,
+	/// Request timeout
```
Comment (dvdplm) [inline http-client/src/client.rs:60]:
```suggestion
	/// Request timeout. Defaults to 60sec.
```

**Review by dvdplm — APPROVED** [review]
LGTM

**Comment by dvdplm** [comment]
Closing in favour of #406

**Pull request closed without merging**